replica/replica.go

```diff
@@ -60,81 +73,153 @@ func New(opts Options, whoami id.Signatory, signatories []id.Signatory, propose
 		proc:         proc,
 		procsAllowed: procsAllowed,
 
-		onTimeoutPropose:   onTimeoutPropose,
-		onTimeoutPrevote:   onTimeoutPrevote,
-		onTimeoutPrecommit: onTimeoutPrecommit,
+		onProposeTimeout:   make(chan timer.Timeout, opts.MessageQueueOpts.MaxCapacity),
+		onPrevoteTimeout:   make(chan timer.Timeout, opts.MessageQueueOpts.MaxCapacity),
+		onPrecommitTimeout: make(chan timer.Timeout, opts.MessageQueueOpts.MaxCapacity),
 
 		onPropose:   make(chan process.Propose, opts.MessageQueueOpts.MaxCapacity),
 		onPrevote:   make(chan process.Prevote, opts.MessageQueueOpts.MaxCapacity),
 		onPrecommit: make(chan process.Precommit, opts.MessageQueueOpts.MaxCapacity),
 		mq:          mq.New(opts.MessageQueueOpts),
+
+		didHandleMessage: didHandleMessage,
 	}
 }
 
+// Run starts the Hyperdrive replica's process
 func (replica *Replica) Run(ctx context.Context) {
 	replica.proc.Start()
-	for {
-		select {
-		case <-ctx.Done():
-			return
 
-		case timeout := <-replica.onTimeoutPropose:
-			replica.proc.OnTimeoutPropose(timeout.Height, timeout.Round)
-		case timeout := <-replica.onTimeoutPrevote:
-			replica.proc.OnTimeoutPrevote(timeout.Height, timeout.Round)
-		case timeout := <-replica.onTimeoutPrecommit:
-			replica.proc.OnTimeoutPrecommit(timeout.Height, timeout.Round)
-
-		case propose := <-replica.onPropose:
-			if !replica.filterHeight(propose.Height) {
-				continue
-			}
-			if !replica.filterFrom(propose.From) {
-				continue
+	isRunning := true
+	for isRunning {
+		func() {
+			defer func() {
+				if replica.didHandleMessage != nil {
+					replica.didHandleMessage()
+				}
+			}()
+
+			select {
+			case <-ctx.Done():
+				isRunning = false
+				return
+
+			case timeout := <-replica.onProposeTimeout:
+				if !replica.filterHeight(timeout.Height) {
+					return
+				}
```

Comment:
These are self-generated messages, so we should not filter by height. The `Process` logic already takes care of timeouts that are from the "wrong" height. 
```suggestion
```